apm-agent-plugins/apm-javalin-plugin/src/main/java/co/elastic/apm/agent/javalin/JavalinInstrumentation.java

```diff
@@ -0,0 +1,163 @@
+                if (name != null) {
+                    transaction.setFrameworkName(FRAMEWORK_NAME);
+                    transaction.setFrameworkVersion(VersionUtils.getVersion(Handler.class, "io.javalin", "javalin"));
+                    transaction.withType("request");
+
+                    name.append(ctx.handlerType().name()).append(" ").append(ctx.endpointHandlerPath());
+
+                    // no need for anonymous handler class names in the transaction
+                    if (!isLambdaHandler) {
+                        name.append(" ").append(handlerClassName);
+                    }
+                }
+            }
+
+            // do not create an own span for JavalinServlet.addHandler, as this not added by the users code and leads to confusion
+            if (handlerClassName.startsWith("io.javalin.http.JavalinServlet")) {
+                return null;
+            }
+
+            // create own span for all handlers including after/before
+            return startSpan(handlerClassName, ctx.handlerType(), isLambdaHandler, ctx.matchedPath(), ctx.method(), ctx.url());
+        }
+
+        @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class, inline = false)
+        public static void onAfterExecute(@Advice.Enter @Nullable Object spanObj,
+                                          @Advice.Argument(0) Object ctxObject,
+                                          @Advice.Thrown @Nullable Throwable t) {
+            if (spanObj != null) {
+                Context ctx = ((Context) ctxObject);
+                final Span span = (Span) spanObj;
+                span.deactivate();
+
+                final CompletableFuture<?> responseFuture = ctx.resultFuture();
+                if (responseFuture == null) {
+                    // sync request
+                    span.captureException(t).end();
+                } else {
+                    // future was set in the handler, so we need to end the span only on future completion
+                    responseFuture.whenComplete((o, futureThrowable) -> span.captureException(futureThrowable).end());
+                }
+            }
+        }
+
+        private static Span startSpan(String handlerClassName, HandlerType handlerType, boolean isLambdaHandler, String matchedPath, String method, String url) {
+            final AbstractSpan<?> parent = tracer.getActive();
+            if (parent == null) {
+                return null;
+            }
+
+            Span span = parent.createSpan().activate();
+            span.appendToName(handlerType.name()).appendToName(" ").appendToName(matchedPath).appendToName(" ");
+            if (isLambdaHandler) {
+                span.appendToName("<Lambda>");
+            } else {
+                span.appendToName(handlerClassName);
+            }
+
+            span.getContext().getHttp().withUrl(url).withMethod(method);
```

Comment:
This context info is only relevant for HTTP client spans, not here.

Reply:
removed